jep/0000/README.adoc

```diff
@@ -0,0 +1,264 @@
+
+Packages should be trackable back to a commit id, but during development the BoM should point to branches or tags, so we would have a BoM as input and a effective BoM as output with all the git refs resolved to specific commit ids.
+
+=== Workflow
+
+Core, libraries and plugins could have a BoM in the master branch.
+This BoM should point to master of core, master of libraries.
+There is an implicit dependency to the current branch of the current git repository.
+
+The BoM should be processed to update the Maven pom on change events.
+
+==== Changes to Core
+
+Developer needs some changes to core or Jenkins internal library or component for downstream use.
+
+* Create PR against core/library.
+** PR builder in core/library will publish the artifacts using the git commit id as part of the identifier, for downstream consumption.
+** Would also publish the result BoM with all git references converted to ids.
+* Create PR against a plugin
+** This PR would include a modified BoM pointing to the PR branch of core/library.
+* PR builder for plugin/library would process the BoM, creating as output:
+** A Jenkins package (WAR and/or Docker image) with the dependencies stated in the BoM.
+** The realized BoM, with all git references converted to ids.
+** Dependencies would not need to be rebuilt as they were already published.
+* Tests run against this output
+* Output artifact is published to repository, for downstream consumption in a way that can be fetched by git commmit id.
+
+==== Changes to Plugins
+
+Same as above but instead of pointing to core/library PR, pointing to master.
+
+==== Continuous Delivery
+
+Changes in master of core and libraries should trigger downstream rebuilds of plugins.
+Given the amount of work involved we recommend doing it in phases, targeting core plugins first, as defined in Essentials.
+
+===== BoM Format
+
+A new yaml format based on https://github.com/jenkins-infra/evergreen/blob/master/essentials.yaml[essentials] and https://github.com/oleg-nenashev/jenkins-war-packager[WAR packager] using the Kubernetes format.
+
+```yaml
+metadata:
+  # labels and annotations are key: value string attributes
+  labels:
+    name: myplugin
+    groupId: ...
+    artifactId: ...
+    version: ...
+  annotations:
+    io.jenkins.x: y
+spec:
+  core:
+    # version OR version + ref (version just to keep Maven happy about version order)
+    ref: master
+    version: 1.0
+  components:
+    - groupId: org.acme
+      artifactId: acme-component
+      ref: master
+      version: 1.0
+  plugins:
+    - groupId: org.acme
+      artifactId: acme-plugin
+      ref: master
+      # version: 1.0
+  environments:
+    # environments get added to the other section when enabled
+    - name: aws
+      plugins:
+        - groupId: org.acme
+          artifactId: acme2-plugin
+          ref: master
+          # version: 1.0
+      components: ...
+  # other sections can be added and ignored by default
+# the realized BoM after refs are resolved
```

Comment:
@carlossg I'm having trouble understanding this part of the format. Is this expected to all be within the same `yaml` file? Or would the "realized BoM" be a separate generated file output?

Reply:
Related to my comment in [JENKINS-50953](https://issues.jenkins-ci.org/browse/JENKINS-50953), I believe the format for "realized" should contain referenceable URLs, This will make this file digestable and immediately usable by the Jenkins Essentials backend tooling.